code/modules/projectiles/projectile.dm

```diff
@@ -674,268 +735,412 @@
+/// Returning TRUE cancels normal animation
+/obj/projectile/proc/move_animate(animate_x, animate_y, animate_time = world.tick_lag, deleting = FALSE)
+	return FALSE
+
+/// Called every projectile loop for homing or alternatively, custom trajectory changes.
+/obj/projectile/proc/process_homing()
 	if(!homing_target)
-		return FALSE
+		return
 	var/datum/point/new_point = RETURN_PRECISE_POINT(homing_target)
 	new_point.x += clamp(homing_offset_x, 1, world.maxx)
 	new_point.y += clamp(homing_offset_y, 1, world.maxy)
 	var/new_angle = closer_angle_difference(angle, angle_between_points(RETURN_PRECISE_POINT(src), new_point))
 	set_angle(angle + clamp(new_angle, -homing_turn_speed, homing_turn_speed))
 
-/obj/projectile/proc/set_homing_target(atom/target)
-	if(!target || (!isturf(target) && !isturf(target.loc)))
-		return FALSE
-	homing = TRUE
-	homing_target = target
-	homing_offset_x = rand(homing_inaccuracy_min, homing_inaccuracy_max)
-	homing_offset_y = rand(homing_inaccuracy_min, homing_inaccuracy_max)
-	if(prob(50))
-		homing_offset_x = -homing_offset_x
-	if(prob(50))
-		homing_offset_y = -homing_offset_y
+/// Attempts to force the projectile to move until the subsystem runs out of processing time, the projectile impacts something or gets frozen by timestop
+/obj/projectile/proc/process_hitscan()
+	if (isnull(movement_vector))
+		qdel(src)
+		return
+
+	while (isturf(loc) && !QDELETED(src))
+		process_movement(ICON_SIZE_ALL, hitscan = TRUE)
+		if (CHECK_TICK || paused || QDELETED(src))
+			return
+
+/// Creates (or wipes clean) list of tracer keypoints and creates a first point.
+/obj/projectile/proc/record_hitscan_start(offset = TRUE)
+	if (isnull(beam_points))
+		beam_points = list()
+	else
+		QDEL_LIST_ASSOC(beam_points)
+		QDEL_NULL(last_point)
+	last_point = RETURN_PRECISE_POINT(src)
+	// If moving, increment its position a bit to prevent it from looking like its coming from firer's ass
+	if (offset && !isnull(movement_vector))
+		last_point.increment(movement_vector.pixel_x * MUZZLE_EFFECT_PIXEL_INCREMENT, movement_vector.pixel_y * MUZZLE_EFFECT_PIXEL_INCREMENT)
+	beam_points[last_point] = null
+
+/// Creates a new keypoint in which the tracer will split
+/obj/projectile/proc/create_hitscan_point(impact = FALSE, tile_center = FALSE, broken_segment = FALSE)
+	var/atom/handle_atom = last_impact_turf || src
+	var/datum/point/new_point = impact ? new /datum/point(handle_atom.x, handle_atom.y, handle_atom.z, impact_x, impact_y) : RETURN_PRECISE_POINT(tile_center ? loc : src)
+	if (!broken_segment)
+		beam_points[last_point] = new_point
+	beam_points[new_point] = null
+	last_point = new_point
+
+/obj/projectile/forceMove(atom/target)
+	if (hitscan && !isnull(beam_points))
+		create_hitscan_point()
+	. = ..()
+	if(!isturf(loc) || !isturf(target) || !z || QDELETED(src) || deletion_queued)
+		return .
+	if (!hitscan || isnull(movement_vector) || isnull(beam_points) || free_hitscan_forceMove)
+		return .
```

Comment:
I am raging against the machine and the machine says there's no reason to return . here unless you want to

Reply:
our standards are stupid